pkg/policy/consumer.go

```diff
@@ -69,81 +111,93 @@ func NewConsumable(id identity.NumericIdentity, lbls *identity.Identity, cache *
 	return consumable
 }
 
+// AddMap adds m to the Consumable's PolicyMaps. This represents
+// the PolicyMap being added for a specific endpoint.
 func (c *Consumable) AddMap(m *policymap.PolicyMap) {
 	c.Mutex.Lock()
 	defer c.Mutex.Unlock()
-	if c.Maps == nil {
-		c.Maps = make(map[int]*policymap.PolicyMap)
+	if c.PolicyMaps == nil {
+		c.PolicyMaps = make(map[int]*policymap.PolicyMap)
 	}
 
 	// Check if map is already associated with this consumable
-	if _, ok := c.Maps[m.Fd]; ok {
+	if _, ok := c.PolicyMaps[m.Fd]; ok {
 		return
 	}
 
 	log.WithFields(logrus.Fields{
 		"policymap":  m,
 		"consumable": c,
 	}).Debug("Adding policy map to consumable")
-	c.Maps[m.Fd] = m
+	c.PolicyMaps[m.Fd] = m
 
 	// Populate the new map with the already established allowed identities from
 	// which ingress traffic is allowed.
 	for ingressIdentity := range c.IngressIdentities {
-		if err := m.AllowIdentity(ingressIdentity.Uint32()); err != nil {
+		if err := m.AllowIdentity(ingressIdentity.Uint32(), Ingress.Uint8()); err != nil {
+			log.WithError(err).Warn("Update of policy map failed")
+		}
+	}
+
+	for egressIdentity := range c.EgressIdentities {
+		if err := m.AllowIdentity(egressIdentity.Uint32(), Egress.Uint8()); err != nil {
 			log.WithError(err).Warn("Update of policy map failed")
 		}
 	}
 }
 
 func (c *Consumable) delete() {
+
 	for ingressIdentity := range c.IngressIdentities {
-		// FIXME: This explicit removal could be removed eventually to
-		// speed things up as the policy map should get deleted anyway
-		if c.wasLastRule(ingressIdentity) {
-			c.removeFromMaps(ingressIdentity)
-		}
+		c.removeFromMaps(ingressIdentity, Ingress)
+	}
+
+	for egressIdentity := range c.EgressIdentities {
+		c.removeFromMaps(egressIdentity, Egress)
 	}
 
 	if c.cache != nil {
 		c.cache.Remove(c)
 	}
 }
 
-func (c *Consumable) RemoveMap(m *policymap.PolicyMap) {
+// RemovePolicyMap removes m from the Consumable's PolicyMaps. This represents
+// the PolicyMap being deleted for a specific endpoint.
+func (c *Consumable) RemovePolicyMap(m *policymap.PolicyMap) {
 	if m != nil {
 		c.Mutex.Lock()
-		delete(c.Maps, m.Fd)
+		delete(c.PolicyMaps, m.Fd)
 		log.WithFields(logrus.Fields{
 			"policymap":  m,
 			"consumable": c,
-			"count":      len(c.Maps),
-		}).Debug("Removing map from consumable")
+			"count":      len(c.PolicyMaps),
+		}).Debug("Removing policy map from consumable")
 
-		// If the last map of the consumable is gone the consumable is no longer
-		// needed and should be removed from the cache and all cross references
-		// must be undone.
-		if len(c.Maps) == 0 {
+		// If there are no more PolicyMaps for this Consumable, then the
+		// Consumable is no longer needed and should be removed from the cache,
+		// and all cross references must be undone.
+		if len(c.PolicyMaps) == 0 {
 			c.delete()
```

Comment:
Is anything locking out others from finding this Consumable and adding themselves to it while this function is executed from another goroutine?

Reply:
The Consumable's mutex is locked on line 115.

Reply:
@raybejjani The clock is ticking....